src/lib.rs

```diff
@@ -126,8 +139,81 @@ pub fn select_coin_knapsack(
-    unimplemented!()
+    let mut target_reached = false;
+    let mut selected_inputs: HashSet<(usize)> = HashSet::new();
+    let mut accumulated_value: u64 = 0;
+    let mut accumulated_weight: u32 = 0;
+    let mut best_set: HashSet<(usize)> = HashSet::new();
+    // Assigning infinity to beginwith
+    let mut best_set_value: u64 = u64::MAX;
+
+    for i in 1..=1000 {
+        for j in 1..=2 {
+            if !target_reached {
+                for &(index, u) in smaller_coins {
+                    //Simulate a coin toss
+                    let mut rng = thread_rng();
+                    let prob = 0.5;
+                    let toss_result: bool = rng.gen_bool(prob);
+                    let selected_element = (index, u);
+                    if (j == 2 && !selected_inputs.contains(&selected_element.0))
+                        || (j == 1 && toss_result)
+                    {
+                        // Including the UTXO in the selected inputs
+                        selected_inputs.insert(selected_element.0);
+                        accumulated_value += u.value;
+                        accumulated_weight += u.weight;
+                        if accumulated_value == adjusted_target {
+                            // Perfect Match, Return the vector selected_inputs
+                            let estimated_fees =
+                                calculate_fee(accumulated_weight, options.target_feerate);
+                            let index_vector: Vec<usize> = selected_inputs.into_iter().collect();
+                            let waste: u64 = calculate_waste(
+                                inputs,
+                                &index_vector,
+                                &options,
+                                accumulated_value,
+                                accumulated_weight,
+                                estimated_fees,
+                            );
+                            return Ok(SelectionOutput {
+                                selected_inputs: index_vector,
+                                waste: WasteMetric(waste),
+                            });
+                        } else if accumulated_value >= adjusted_target {
+                            target_reached = true;
+                            if accumulated_value < best_set_value {
+                                // New best_set found
+                                best_set_value = accumulated_value;
+                                best_set.clone_from(&selected_inputs);
+                                // Removing the last UTXO that raised selection_sum above adjusted_target to try to find a smaller set
+                                selected_inputs.remove(&selected_element.0);
+                                accumulated_value -= u.value;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Best set of UTXOs after 1000 trials
```

Comment:
After the end of for loop there are two possible cases.
1. It is not possible to get a selection with smaller inputs, not exactly Insufficient funds so a different error should be returned
2. A selection which is not an exact match in that case Ok(selection) should be returned

Reply:
1. Shall we add to the SelectionError enum ?
2. Doesn't this 
```
let estimated_fees = calculate_fee(accumulated_weight, options.target_feerate);
    let index_vector: Vec<usize> = selected_inputs.into_iter().collect();
    let waste: u64 = calculate_waste(
        inputs,
        &index_vector,
        &options,
        accumulated_value,
        accumulated_weight,
        estimated_fees,
    );
    Ok(SelectionOutput {
        selected_inputs: index_vector,
        waste: WasteMetric(waste),
      ##}) 
```
handle the case where we end up with a selection of input values > target value ?

Reply:
When the the target is not reached return `NoSolutionsFound` error